### ISSUE TITLE
Plan to update cluster hosts? (proposal looking for feedback)

### ISSUE THREAD

**Issue by TJM** [albatrossflavour/puppet_os_patching#31]
We are looking to implement this type of module to update our systems with automation, however we have several "clusters" where at a minimum, we want to lump together a "set" of machines to update "one at a time" with maybe 30-60 seconds wait after each reboot before starting the next one. However if we go to that level, how hard would it be to have some sort of "health check" that has to pass before the next machine in the group can start? One such health check, for elastic search for example, would be something like..

```
curl ${AUTH} http://${HOSTNAME}:9200/_cat/health | grep -q green
```

Basically the concept would be that once the target host is able to be connected to and can query it's own cluster status to say "green" then it passes the health check and the next cluster node can start patching.

So, the default health check could be something like `sleep 5` with the ability to customize it with a "script" (content/source).

We could potentially even have some "example" health checks for common cluster types (redis? elastic?) There could maybe even be something to patch docker hosts, but that might be better handled by a docker specific "task" to drain a node, for example.

Anyhow, this is a proposal for something that we "want" but we don't really have a firm grasp on what exactly we want or what the capabilities are, so I figured I would start here looking for feedback.

Thanks in advance,
Tommy

**Comment by rnelson0** [comment]
Brainstorming here, so bear with the stream of consciousness...

The current `patch_window` setting just accepts an arbitrary string, so you could manually assign different nodes to different groups like `ClusterMembersA`, `ClusterMembersB`, etc. and schedule your jobs appropriately to run these groups in sequence. You could semi-automate assignment through some classifier rules, e.g. VMs with this tag get put in `ClusterMembersA`, this other tag in `ClusterMembersB`, etc.

I do not think it would be appropriate to move classification into a puppet module, but the previous solution is still fairly manual (or requires additional automation elsewhere). What could be appropriate in the task is error handling + delay. Currently, the task enumerates all nodes and fires off a job for each node with an independent job ID. A delay would not help, since jobs are simultaneous. However, a Plan could be used to orchestrate the tasks. The Plan could accept a few arguments:
* delay (integer, in seconds, defines the time between a run and external health checks/next task, default `0`)
* simultaneous_tasks (integer, default `1`)
* external_health_check (string array - but REST? script? where would it run? default `undef`)
* abort_on_error (boolean, when true either a failure in the task results or any external health check stops the plan)
* patch_window (string, default undef but required)

This would allow the flow to be:
* Schedule plan, providing mandatory node selection and arg patch_window, plus optional args
* Plan crafts the list of tasks and orders them in serial (user would have to craft the correct nodelist/query to ensure proper ordering), with the order
  * Task for Node A
  * Verify success of Node A
  * Delay
  * Perform any external health checks of Node A
  * If task or check failure and abort_on_error is true, end the plan
  * Otherwise, repeat for subsequent nodes

I've not used Plans before, this is based off my reading of the docs and general understanding of the concept, so it might need some tweaking or perhaps be flat our wrong. If it is not, does this sound like the flow you're looking for @TJM?

**Comment by albatrossflavour** [comment]
I had thought about how we can do pre-checks and validation prior to doing any patching, but each solution had big holes in it.  The closest I came to something I liked was to have an option to specify a fact to check and a value to check for.  This way you can have an independent health check running and the patching just queries that to see if it's allowed to proceed.

Plans are certainly the other option however they're very much a separate setup at the moment.  I'm going to include some plan examples for advanced users, but I don't want to directly tie the use of the module to using bolt.

Bolt does give you the advantage of having concurrency options, which tasks is still (for some reason) missing.

What do you think @TJM ?

**Comment by TJM** [comment]
A pre-check might work, in some cases, but would be difficult for what I have in mind, because you would need to make it aware of what node patched previously to check it. I suppose that would also depend on what "kind" of cluster it was, whether it could check to see if there were some minimum number of nodes or whatever. I think a pre-check would be a good "first step" on the way to completely automating patching. :)

I do not think *facts* will work for health checking specifically because they are slow to update, and the conditions/decisions based on fact values are usually made at the time the job is created rather than at execution time. They can certainly be used to logically group cluster nodes, either getting all nodes of a particular cluster (and setting concurrency -> 1?) or getting "inactive nodes" or whatever. I suppose that you could execute `facter -p ${healthcheck_fact}` as a pre-check, but then you would have to parse it (` | ${healthcheck_parse_cmd}` ( maybe something like `grep -q 'status => "ok"'`). It would need to take care of timeouts, retries, etc. Once the ability to run a healthcheck with (max)retries/timeouts, the script/command becomes arbitrary, so I wouldn't limit it to using facter.

An example plan would be a good first step, maybe we could use the puppetlabs-exec module to run the health check? I'd like to see it integrated into the module at some point. I didn't realize that plans were "bolt only" (?). I see a "Plan" tab in the 2018.1.4 UI under jobs, but dont see a Plan section in the left menu, maybe coming soon? :)

Puppet tasks run concurrently, but we aren't able to configure that concurrency, right? I think bolt tries to do 100 concurrent nodes, which in my experience breaks stuff. It would certainly be nice to have concurrency configuration.

~tommy

**Comment by albatrossflavour** [comment]
I'm going to park this one for now in the hope that October brings us more news on tasks/bolt enhancements which we can use to improve this.  I don't think updating the current modules features to deal with the dependancy scenario is the right move, but I'll look at adding in an example bolt file and some info on how to use it.
